### ISSUE TITLE
Scorm (no rating) not attempting

### ISSUE THREAD

**Issue by andre-dietrich** [overhangio/openedx-scorm-xblock#43]
Hi and thanks for the awesome project. I wanted to upload SCORM content, which in principal works, but I tried lots of different configurations, I cannot rate the course. It is always in state `not attempted` as you can see from the screenshots.

![image](https://github.com/overhangio/openedx-scorm-xblock/assets/3089101/997d3494-641f-4c51-9fa1-2192b65510ac)

I cannot read the `cmi.mode`, which should be "normal" to rate the course, however I always get an empty string back. When I ignore this setting and store the results via the SCORM API, the values are stored persistent, but not rated. Is there something fundamental that I am missing? I also get the same results by using the standard SCORM-examples from here:

https://scorm.com/scorm-explained/technical-scorm/golf-examples/

THX for any hints or suggestions ...

**Comment by andre-dietrich** [comment]
According to the scorm-standard defined:

https://scorm.com/scorm-explained/technical-scorm/run-time/

> ## Entry, Mode and Credit
>
> The `cmi.entry` (`cmi.core.entry`), `cmi.mode` (`cmi.core.lesson_mode`) and `cmi.credit` (`cmi.core.credit`) data model elements provide the SCO with some context it can use to provide the learner with the optimal experience.
>
> __`cmi.entry`__ indicates whether the user is starting the SCO for the first time or if it is resuming a prior attempt. If using bookmarking, the SCO might use this value to prompt the user to start from the beginning or from the point at which the previous attempt ended.
>
> __`cmi.mode`__ indicates whether the learner is launching this SCO: normally – in a “live” training session; in browse mode – the user is browsing a catalog of training and wants to “preview” this course; or, in review mode – the user has already completed the SCO and is coming back to review the material. In browse mode, the SCO author should consider altering the SCOs behavior to provide more free form navigation, provide an overview or map of the course and possibly hide assessments. In review mode, the SCO author should similarly consider allowing the leaner complete freedom of navigation. In review mode, the learner might also be able to freely navigate any tests and browse a list of correct answers. When launched in review mode, a SCO should be careful not to change the learner’s status or reset the score.
>
> __`cmi.credit`__ indicates whether or not this SCO is being attempt for credit, or whether or not it “counts”. Similar to browse mode, if a SCO is launched for no credit, the behavior should probably be altered.

I think the `cmi.mode` or `cmi.core.lesson_mode` could be easily taken from the LMS, and are defined by the state the user is in, either it is in preview-mode or active:

![image](https://github.com/overhangio/openedx-scorm-xblock/assets/3089101/fc2b90fa-b4a1-4312-8eb3-9b121f3b260f)

so that these settings match the current state.

The `cmi.mode` is actually only read only.

**Comment by regisb** [comment]
Hi @andre-dietrich! I'm not sure I understand your question. The "not attempted" message and grade are displayed on the basis of the values "cmi.completion_status" and "cmi.success_status". These seem unrelated to the "cmi.mode" and "cmi.core.lesson_mode" flags you are mentioning.

> I cannot rate the course

What do you mean exactly by that? I do not know what "rating" means in the context of scorm.

**Comment by andre-dietrich** [comment]
Hi, to my knowledge, the mode has to be set to differentiate between a scored course, a preview or if another person is reviewing the course or the results of another user:

![image](https://github.com/overhangio/openedx-scorm-xblock/assets/3089101/78e69992-66cc-4f74-9533-d1b1a6dbd051)

https://scorm.com/scorm-explained/technical-scorm/run-time/

I tried this also with the official SCORM-golf examples, you can download them here:

https://scorm.com/scorm-explained/technical-scorm/golf-examples/

In all cases the course could not be loaded, because the expected `cmi.mode` was not set. I just tried out edX for the first time, but I think these "view this course as" could be matched to the `cmi.mode`.

**Comment by regisb** [comment]
> What do you mean exactly by that? I do not know what "rating" means in the context of scorm.

Please help me better understand your problem by answering my [question above](https://github.com/overhangio/openedx-scorm-xblock/issues/43#issuecomment-1729619004).

**Comment by regisb** [comment]
Right, I think I better understand your issue.

But yes, I did run some of the official SCORM packages. Here's a screenshot of an xblock for which I attempted the MCQ:

![Screenshot from 2023-09-22 10-24-50](https://github.com/overhangio/openedx-scorm-xblock/assets/44319/cb3a48ed-656c-4c49-a73d-39dfb37bc8a3)

This is the "RuntimeBasicCalls_SCORM20043rdEdition.zip" golf example.

- As you can see in the top left, the MCQ attempt was correctly graded.
- At the bottom you can see the actual data that is stored for the user. It does not include "cmi.mode".

You can try it out on the demo server: https://apps.demo.openedx.overhang.io/learning/course/course-v1:edX+DemoX+Demo_Course/block-v1:edX+DemoX+Demo_Course+type@sequential+block@528a58b1a9a24f4a8d665e00098c7e90/block-v1:edX+DemoX+Demo_Course+type@vertical+block@1c1aae685dfb47fdadd9dd73ca8df5f2

Exactly what zip file does not work for you?

**Comment by andre-dietrich** [comment]
Okay, now I see the problem ... The `cmi.mode` is a read-only value, when it is set to browse or review by the LMS, it indicates to the SCORM-package that new results should not be stored (at least that is how I interpret this). Thus in Moodle and other LMS this value is set and probably also writing-state back is blocked in such a case.

> `cmi.mode` (“browse”, “normal”, “review”, RO) Identifies one of three possible modes in which the SCO may be presented to the learner

Thus, our export retrieves this value from the LMS, such as visiting the course only for preview, will not result in any storage to the backend. If this value is ignored, and the backend does not protect writing, this will work, but not as we expect it to be...

That is, why I thought, the "audit" setting can be used to match this, whether a student has enrolled the course, thus cmi.mode = "normal", or staff and others are only testing "browse", etc.

I hope this makes sense?

**Comment by andre-dietrich** [comment]
By the way, this is a course I want to upload to edX

[course.zip](https://github.com/overhangio/openedx-scorm-xblock/files/12699654/output.zip)

**Comment by regisb** [comment]
> I hope this makes sense?

Nope, not for me :sweat: It's super hard for me to understand your explanation if I don't understand what is your problem in the first place. You can help me by answering my questions. In particular:

> Exactly what zip file does not work for you?

The "output.zip" file you uploaded does not look like a standard golf example from the SCORM website. Do you have a standard example that does not work with this xblock?

**Comment by andre-dietrich** [comment]
The SCORM 1.2 works with Moodle, 2004 with for example ILIAS, here is the example for 1.2:

[output.zip](https://github.com/overhangio/openedx-scorm-xblock/files/12700605/output.zip)

As you can see from the video, Moodle sets the value `cmi.mode` in preview mode and normal, if I enroll the course:

https://www.youtube.com/watch?v=apV2BW-MPho

(video is atm processed)

is it possible for you, to set the `cmi.mode` based on the Audit-mode, to "normal", "browse" or "review"?

![image](https://github.com/overhangio/openedx-scorm-xblock/assets/3089101/cd165dda-038b-4ae2-95b3-a582484d04dc)

**Comment by regisb** [comment]
> is it possible for you, to set the cmi.mode based on the Audit-mode, to "normal", "browse" or "review"?

I have done this in my dev environment. But I do not know whether what I am doing is is "right" because all you gave me is a large, custom SCORM unit. Setting the "cmi.mode" and "cmi.core.lesson_mode" to "normal" does not change the "attempted" status.

Please do read my comments. In particular:

> The "output.zip" file you uploaded does not look like a standard golf example from the SCORM website. Do you have a standard example that does not work with this xblock?

**Comment by ziafazal** [comment]
We are unable to reproduce this issue. @andre-dietrich Could you please list down steps to reproduce this one?

**Comment by andre-dietrich** [comment]
The problem was, that SCORM has different modes for preview or review in which it can operate, these are not set by your plugin. I now integrated a special case, if `cmi.mode` is not set the course and all results are stored in the LMS ...

**Comment by rimsha-zaib** [comment]
I tried with multiple standard SCORM-examples, the MCQ attempt was correctly graded. As below,

![Screenshot 2024-01-05 at 12 36 35 PM](https://github.com/overhangio/openedx-scorm-xblock/assets/137502034/77aa0fa3-ce21-4c77-9bd0-8dafc4ea941f)

**Comment by andre-dietrich** [comment]
Yes, I know that it works and grading works also ... but you can set a course to __review__ or __preview__ mode, without any scoring at all, where others can try it out before it gets published ... This "cmi.mode" can be passed to the SCORM-package. In this case no rating will happen and no scores or states will be stored ...

**Comment by Danyal-Faheem** [comment]
Closing this issue as there has already been a fix regarding this in #64. Furthermore, the course provided by the issue reporter is not a standard scorm golf course, so there are no promises non standard courses will work with this xblock.
